aten/src/ATen/core/ivalue.h

```diff
@@ -367,6 +368,28 @@ struct CAFFE2_API IValue final {
     throw std::runtime_error("IValue is not a Scalar");
   }
 
+  // Device
+  IValue(c10::Device d)
+  : tag(Tag::Device), is_intrusive_ptr(false) {
+    payload.as_device.type = d.type();
+    payload.as_device.index = d.index();
+  }
+  bool isDevice() const { return Tag::Device == tag; }
+  c10::Device toDevice() const {
+    AT_ASSERT(isDevice());
+    return c10::Device(payload.as_device.type, payload.as_device.index);
+  }
+
+  // ScalarType
+  at::ScalarType toScalarType() const {
+    return static_cast<at::ScalarType>(toInt());
```

Comment:
assert isDevice here I think.

Reply:
`toInt` does the assert, and that is all we know (until we a add first-class ScalarType support)